WKWebView does not respect `-[UIScrollView allowsKeyboardScrolling]`

rr-codes opened the pull request:
#### ff056c129d9c43a15b6b2a52e9ab874d8b50eaba
<pre>
WKWebView does not respect `-[UIScrollView allowsKeyboardScrolling]`
<a href="https://bugs.webkit.org/show_bug.cgi?id=263164">https://bugs.webkit.org/show_bug.cgi?id=263164</a>
rdar://115803459

Reviewed by Wenson Hsieh.

When determining if a keyboard scroll may occur, consult the value of `-[UIScrollView allowsKeyboardScrolling]`,
a new API introduced in iOS 17.

The default value of this property is `NO` on iPhone, and `YES` on iPad, Catalyst, and Vision. As such, the
property must initially be set to `YES` in WKScrollView to maintain its existing behavior.

Also adds an API test for this behavior.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView isScrollableForKeyboardScrollViewAnimator:]):
* Source/WebKit/UIProcess/ios/WKScrollView.mm:
(-[WKScrollView initWithFrame:]):
* Tools/TestWebKitAPI/Tests/ios/WKScrollViewTests.mm:
(TEST):

Canonical link: <a href="https://commits.webkit.org/269378@main">https://commits.webkit.org/269378@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/19d939c100b137fcda08f3e04820a0f8c5ce3621

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22315 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22514 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23390 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24215 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20653 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22565 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26776 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22844 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21662 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22551 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/22251 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19370 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25070 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/19301 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20225 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26472 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/19407 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20318 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20456 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24332 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/21683 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20984 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17779 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/25734 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/20442 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6057 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5333 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24672 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/27009 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20980 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5884 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -6722,6 +6722,13 @@ - (BOOL)isScrollableForKeyboardScrollViewAnimator:(WKKeyboardScrollViewAnimator
 
+#if HAVE(UISCROLLVIEW_ALLOWS_KEYBOARD_SCROLLING)
+    // On iPhone, the default value is `NO`.
+    // On iPad, Catalyst, and Vision, the default value is `YES`.
```
This comment should be removed now.

pxlcoder commented on Source/WebKit/UIProcess/ios/WKScrollView.mm:
```diff
@@ -173,6 +173,13 @@ - (id)initWithFrame:(CGRect)frame
 
+#if HAVE(UISCROLLVIEW_ALLOWS_KEYBOARD_SCROLLING)
+    // In iOS 17, the default value of `-[UIScrollView allowsKeyboardScrolling]` is `NO`.
+    // To maintain existing behavior of WKScrollView, this property must be iniitally set to `YES`.
```
`iniitally`

pxlcoder commented on Tools/TestWebKitAPI/Tests/ios/WKScrollViewTests.mm:
```diff
@@ -296,4 +297,54 @@ static void traverseLayerTree(CALayer *layer, void(^block)(CALayer *))
+
+        auto secondWebEvent = adoptNS([[WebEvent alloc] initWithKeyEventType:WebEventKeyUp timeStamp:CFAbsoluteTimeGetCurrent() characters:@" " charactersIgnoringModifiers:@" " modifiers:0 isRepeating:NO withFlags:0 withInputManagerHint:nil keyCode:0 isTabKey:NO]);
+
+        [contentView handleKeyWebEvent:firstWebEvent.get() withCompletionHandler:^(WebEvent *theEvwnt, BOOL wasHandled) {
```
`theEvwnt`

pxlcoder reviewed (commented):
Commit message needs to be updated to reflect the newly changed file.

rr-codes committed:
WKWebView does not respect `-[UIScrollView allowsKeyboardScrolling]`
https://bugs.webkit.org/show_bug.cgi?id=263164
rdar://115803459

Reviewed by Wenson Hsieh.

When determining if a keyboard scroll may occur, consult the value of `-[UIScrollView allowsKeyboardScrolling]`,
a new API introduced in iOS 17.

The default value of this property is `NO` on iPhone, and `YES` on iPad, Catalyst, and Vision. As such, the
property must initially be set to `YES` in WKScrollView to maintain its existing behavior.

Also adds an API test for this behavior.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView isScrollableForKeyboardScrollViewAnimator:]):
* Source/WebKit/UIProcess/ios/WKScrollView.mm:
(-[WKScrollView initWithFrame:]):
* Tools/TestWebKitAPI/Tests/ios/WKScrollViewTests.mm:
(TEST):

Canonical link: https://commits.webkit.org/269378@main
